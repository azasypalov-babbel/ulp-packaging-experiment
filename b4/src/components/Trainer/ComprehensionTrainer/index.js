export { default as default } from './ComprehensionTrainerContainer';
