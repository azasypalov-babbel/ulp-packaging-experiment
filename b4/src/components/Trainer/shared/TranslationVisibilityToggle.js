import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import TranslationToggle from '../../shared/TranslationToggle';

const StyledToggleWrapper = styled.div`
  position: sticky;
  padding: 0.75rem 0; 
  top: 0.75rem;
  left: 0.75rem;
  float: left;

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    left: 1.5rem;
  }
`;

const TranslationVisibilityToggle = ({ onClick, translationVisibility }) => (
  <>
    {translationVisibility === 'partial' &&
      <StyledToggleWrapper>
        <TranslationToggle onClick={onClick} />
      </StyledToggleWrapper>
    }
  </>
);

TranslationVisibilityToggle.propTypes = {
  translationVisibility: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired
};

export default TranslationVisibilityToggle;
