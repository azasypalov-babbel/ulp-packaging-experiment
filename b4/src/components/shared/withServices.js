import React, { Component } from 'react';
import { getDisplayName } from '../getDisplayName';

export const ServiceContext = React.createContext({});

const pick = (obj, arr) =>
  arr.reduce((acc, curr) => (curr in obj && (acc[curr] = obj[curr]), acc), {});

export const withServices = (serviceNames) => (WrappedComponent) => {
  const WithServices = (props) => (
    <ServiceContext.Consumer>
      {(services) => (
        <WrappedComponent
          {...pick(services, serviceNames)}
          {...props}
        />
      )}
    </ServiceContext.Consumer>
  );

  WithServices.displayName = `WithServices(${getDisplayName(WrappedComponent)})`;

  return WithServices;
};

export const withServicesProvider = (selectServices) => (WrappedComponent) => {
  class ServicesProvider extends Component {
    constructor(props) {
      super(props);
      this.services = selectServices();
    }
    render() {
      return (
        <ServiceContext.Provider value={this.services}>
          <WrappedComponent {...this.props} />
        </ServiceContext.Provider>
      );
    }
  }

  ServicesProvider.displayName = `ServicesProvider(${getDisplayName(WrappedComponent)})`;

  return ServicesProvider;
};
