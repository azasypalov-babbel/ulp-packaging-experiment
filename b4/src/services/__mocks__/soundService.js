const realSoundService = jest.requireActual('../soundService');
export const PlayStates = realSoundService.PlayStates;
export const checks = realSoundService.checks;

const mockInstances = new Map();

const createMockSoundController = jest.fn((url) => {
  const mockListeners = new Set();
  let playState = PlayStates.INITIAL;
  const controller = {
    getPlaybackRate: jest.fn(() => 1.0),
    getState: jest.fn(() => playState),
    getDuration: jest.fn(() => 0.0),
    load: jest.fn(() => Promise.resolve()),
    cleanup: jest.fn(),
    play: jest.fn(),
    pause: jest.fn(),
    stop: jest.fn(),
    reset: jest.fn(() => {
      playState = PlayStates.INITIAL;
    }),
    subscribe: jest.fn((listener) => {
      mockListeners.add(listener);
      return () => {
        mockListeners.delete(listener);
      };
    }),
    // Mock specific helper methods
    simulate: (topic, msg) => {
      const stateMap = {
        playinitiated: PlayStates.INITIATED,
        play: PlayStates.PLAYING,
        pause: PlayStates.PAUSED,
        stop: PlayStates.STOPPED,
        end: PlayStates.COMPLETED,
        playerror: PlayStates.FAILED
      };
      if (stateMap[topic]) {
        playState = stateMap[topic];
      }
      mockListeners.forEach((listener) => listener({ event: topic, msg }));
    }
  };

  mockInstances.set(url, controller);

  return controller;
});

const mockSoundService = {
  preload: jest.fn((url) => {
    if (!mockInstances.has(url)) createMockSoundController(url);
  }),
  subscribe: jest.fn(),
  stop: jest.fn(),
  play: jest.fn(),
  isPlaying: jest.fn(
    (instance) => [PlayStates.INITIATED, PlayStates.PLAYING].includes(instance.getState())
  ),
  hasPlayed: jest.fn(
    (instance) => ![
      PlayStates.INITIATED,
      PlayStates.PLAYING,
      PlayStates.STOPPED
    ].includes(instance.getState())
  ),
  getPlayingInstances: jest.fn(() => []),
  getInstance: jest.fn((url) => {
    mockSoundService.preload(url);
    return mockInstances.get(url);
  })
};

afterEach(() => {
  mockInstances.clear();
});

export default mockSoundService;
