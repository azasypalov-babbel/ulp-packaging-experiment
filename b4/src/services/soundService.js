import { Howl, Howler } from 'howler';
import rollbar from './rollbarService';

export const PlayStates = {
  FAILED: 'failed',
  STOPPED: 'stopped',
  INITIAL: 'initial',
  INITIATED: 'initiated',
  PLAYING: 'playing',
  PAUSED: 'paused',
  COMPLETED: 'completed'
};

export const checks = {
  isPlaying(playState, includingInitiated = true) {
    return [
      PlayStates.PLAYING,
      includingInitiated ? PlayStates.INITIATED : PlayStates.PLAYING
    ].includes(playState);
  },
  hasPlayed(playState, notPaused = false) {
    return !checks.isPlaying(playState) && ![
      PlayStates.INITIAL,
      notPaused ? PlayStates.PAUSED : PlayStates.INITIAL
    ].includes(playState);
  },
  hasFailed(playState) {
    return playState === PlayStates.FAILED;
  },
  isInterrupted(playState) {
    return checks.hasPlayed(playState) && playState !== PlayStates.COMPLETED;
  }
};

const getUrl = (soundData) => soundData?.src ? soundData.src : soundData;

export const createSoundService = () => {
  const createPubSub = () => {
    const listeners = [];
    return {
      subscribe(listener) {
        if (!listeners.includes(listener)) {
          listeners.push(listener);
        }
        return () => {
          const index = listeners.indexOf(listener);
          if (index !== -1) {
            listeners.splice(index, 1);
          }
        };
      },
      publish(...args) {
        listeners.forEach((listener) => listener(...args));
      },
      unsubscribe() {
        listeners.splice(0, listeners.length);
      }
    };
  };
  const instances = {};
  const globalPubSub = createPubSub();
  const createSoundController = (soundData, service) => {
    let playState = PlayStates.INITIAL;
    const url = getUrl(soundData);
    const howl = !url ? null : new Howl({
      src: url,
      html5: true,
      preload: true,
      ...(soundData && typeof soundData === 'object' ? soundData : {})
    });
    const pubSub = createPubSub();
    const createFreezeLogger = () => {
      let timers = {};
      const unsub = pubSub.subscribe(() => {
        const newState = playState;
        Object.entries(timers).forEach(([stateKey, timer]) => {
          if (newState !== stateKey) {
            clearTimeout(timer);
            delete timers[stateKey];
          }
        });
      });
      return {
        track({ fallbackState, publishEvent = 'playerror', timeout = 3000 }) {
          const currentPlayState = playState;
          clearTimeout(timers[currentPlayState]);
          timers[currentPlayState] = setTimeout(() => {
            rollbar.warning(`Sound has frozen and was skipped`, {
              url,
              state: currentPlayState,
              howlerState: howl.state()
            });
            playState = fallbackState;
            pubSub.publish({ event: publishEvent });
            delete timers[currentPlayState];
          }, timeout);
        },
        unwire() {
          unsub();
          Object.values(timers).forEach((timer) => {
            clearTimeout(timer);
          });
          timers = {};
        }
      }
    }
    const freezeLogger = createFreezeLogger();
    let loadingPromise = null;
    const unsubFromGlobal = pubSub.subscribe((topic, value) => {
      globalPubSub.publish(instances[url], topic, value);
    });
    const controller = {
      getPlaybackRate() {
        if (howl === null || ['unloaded', 'loading'].includes(howl.state())) return 1;
        return howl.rate();
      },
      getState() {
        return playState;
      },
      getDuration() {
        if (howl === null || ['unloaded', 'loading'].includes(howl.state())) return 0;
        return (howl.duration() * 1000) / controller.getPlaybackRate();
      },
      load(retries = 0) {
        if (loadingPromise === null) {
          loadingPromise = howl === null ? Promise.reject() : new Promise((resolve, reject) => {
            if (howl.state() !== 'loading') howl.load();
            howl.once('load', resolve);
            howl.once('loaderror', () => {
              if (retries === 0) {
                howl.unload();
                reject();
                loadingPromise = null;
              } else {
                controller.load(retries - 1).then(resolve, reject);
              }
            });
          });
        }
        return loadingPromise;
      },
      reset() {
        if (playState === PlayStates.PAUSED) howl?.seek(0);
        playState = PlayStates.INITIAL;
      },
      cleanup() {
        if (howl !== null) {
          howl.off('play');
          howl.off('end');
          howl.off('pause');
          howl.off('stop');
          howl.off('playerror');
          howl.off('loaderror');
        }
        loadingPromise = null;
        unsubFromGlobal();
        pubSub.unsubscribe();
        freezeLogger.unwire();
        playState = PlayStates.INITIAL;
        delete instances[url];
      },
      play({ reset = false, stopOtherSounds = true, playbackRate = null, skipOnFreeze = false } = {}) {
        if (howl === null) {
          // to update the state
          playState = PlayStates.INITIATED;
          pubSub.publish({ event: 'playinitiated' });
          setTimeout(() => {
            playState = PlayStates.FAILED;
            pubSub.publish({ event: 'playerror' });
          });
          return;
        }
        if (stopOtherSounds) service.stop({ exceptUrls: [url] });
        if (reset && playState !== PlayStates.INITIATED) howl.seek(0);
        if (reset || ![PlayStates.PLAYING, PlayStates.INITIATED].includes(playState)) {
          playState = PlayStates.INITIATED;
          pubSub.publish({ event: 'playinitiated' });
          if (skipOnFreeze) freezeLogger.track({ fallbackState: PlayStates.FAILED });
          if (playbackRate !== null) howl.rate(playbackRate);
          if (howl.state() === 'unloaded') {
            controller.load().finally(() => {
              if (playState === PlayStates.INITIATED) howl.play();
            });
          }
          else howl.play();
        }
      },
      pause() {
        if (howl !== null) howl.pause();
        else {
          setTimeout(() => {
            pubSub.publish({ event: 'playerror' });
          });
        }
      },
      stop() {
        if (howl !== null) howl.stop();
        else {
          setTimeout(() => {
            pubSub.publish({ event: 'playerror' });
          });
        }
      },
      unload() {
        howl?.unload();
      },
      subscribe: pubSub.subscribe
    };
    if (howl !== null) {
      howl.on('play', (_, msg) => {
        playState = PlayStates.PLAYING;
        pubSub.publish({ event: 'play', msg });
      });
      howl.on('pause', (_, msg) => {
        playState = PlayStates.PAUSED;
        pubSub.publish({ event: 'pause', msg });
      });
      howl.on('stop', (_, msg) => {
        if (playState !== PlayStates.INITIAL) playState = PlayStates.STOPPED;
        pubSub.publish({ event: 'stop', msg });
      });
      howl.on('end', (_, msg) => {
        playState = PlayStates.COMPLETED;
        pubSub.publish({ event: 'end', msg });
      });
      howl.on('loaderror', (_, msg) => {
        if (playState !== PlayStates.INITIAL) playState = PlayStates.FAILED;
        pubSub.publish({ event: 'loaderror', msg });
        loadingPromise = null;
        controller.unload();
      });
      howl.on('playerror', (_, msg) => {
        playState = PlayStates.FAILED;
        pubSub.publish({ event: 'playerror', msg });
      });
    }
    return controller;
  };
  const soundService = {
    preload: (soundData, { retriesOnFail = 0 } = {}) => {
      const url = getUrl(soundData);
      if (!instances[url]) {
        instances[url] = createSoundController(soundData, soundService);
      }
      instances[url].load(retriesOnFail).catch(() => {
        instances[url].unload();
      });
    },
    subscribe: globalPubSub.subscribe,
    play(url, options = {}) {
      const instance = soundService.getInstance(url);
      instance.play(options);
      return instance;
    },
    stop({ exceptUrls = [] } = {}) {
      Object.entries(instances).forEach(([url, instance]) => {
        if (!exceptUrls.includes(url) && soundService.isPlaying(instance)) {
          instance.stop();
        }
      });
    },
    isPlaying(soundDataOrInstance) {
      if (!soundDataOrInstance) return false;
      const url = getUrl(soundDataOrInstance);
      const instance = typeof url === 'string' ? instances[url] : soundDataOrInstance;
      return checks.isPlaying(instance.getState());
    },
    hasPlayed(urlOrInstance) {
      if (!urlOrInstance) return false;
      const url = getUrl(urlOrInstance);
      const instance = typeof url === 'string' ? instances[url] : urlOrInstance;
      return checks.hasPlayed(instance.getState());
    },
    getPlayingInstances() {
      return Object.values(instances).filter(soundService.isPlaying);
    },
    getInstance(soundData) {
      const url = getUrl(soundData);
      if (!instances[url]) soundService.preload(soundData);
      return instances[url];
    }
  };

  document.addEventListener('DOMContentLoaded', () => {
    // By setting volume & calling unlockAudio, howler will
    // try to unlock the audio permissions before any sounds are even loaded.
    Howler.volume();
    // eslint-disable-next-line no-underscore-dangle
    Howler._unlockAudio();
  });

  return soundService;
};

export default createSoundService();
