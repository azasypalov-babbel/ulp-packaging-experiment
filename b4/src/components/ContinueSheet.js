import React from 'react';
import PropTypes from 'prop-types';

import Button from './shared/Button/Button';
import Sheet, { FillAvailableSpace, PreserveSize } from './shared/Sheet';
import withTranslations from './shared/withTranslations';
import { compose } from 'redux';
import { renderInBottomLayout } from './shared/BottomLayout';
import styled from 'styled-components';
import throttle from '../lib/throttle';

const StyledButtonWrapper = styled.div`
  width: 100%;

  & > button {
    width: 100%;

    @media (min-width: 620px) {
      width: auto;
    }
  }
`;


export const ContinueSheet = ({ translations, onClick, children }) => (
  <Sheet onClick={(event) => event.stopPropagation()} data-selector="continue-sheet">
    <FillAvailableSpace>
      {children}
    </FillAvailableSpace>
    <PreserveSize>
      <StyledButtonWrapper>
        <Button
          primary
          data-selector="continue-button"
          onClick={throttle(onClick, 300)}
          listenToKey="Enter"
        >
          {translations.buttonText}
        </Button>
      </StyledButtonWrapper>
    </PreserveSize>
  </Sheet>
);

ContinueSheet.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    buttonText: PropTypes.string.isRequired
  }).isRequired
};

const getTranslations = (translate) => ({
  buttonText: translate('listening_trainer.button.continue')
});

export default compose(
  withTranslations(getTranslations),
  renderInBottomLayout
)(ContinueSheet);
