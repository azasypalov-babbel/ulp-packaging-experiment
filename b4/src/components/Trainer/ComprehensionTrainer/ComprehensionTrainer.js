import PropTypes from 'prop-types';
import React, { useContext, useMemo } from 'react';
import styled, { keyframes } from 'styled-components';

import Player from '../../Player/Player';
import withTranslations from '../../shared/withTranslations';
import StyledTrainerLayout from '../../TrainerLayout';
import { ServiceContext } from '../../shared/withServices';
import { ComprehensionTrainerTitle, StyledDescriptionWrapper, StyledWrapper } from './styles';
import ContextualInfo from '../../shared/ContextualInfo';
import { TextContent } from './TextContent';

const getTranslations = (translate) => ({
  trainerTitle: {
    text: translate('b3.comprehension.text_title'),
    audio: translate('b3.comprehension.audio_title')
  }
});

const fadeInAnimation = keyframes`
  from {
    height: 100%;
    opacity: 0.5;
  }
  to {
    height: 100%;
    opacity: 1;
  }
`

const TrainerBody = styled.div`
  animation: ${fadeInAnimation} 400ms ease-out 1.5s 1 forwards;
  opacity: 0;
  height: 0;
`;

export const ComprehensionTrainer = ({
  onAllItemsComplete,
  dictate,
  items,
  title,
  translations,
  description,
  translationsVisible
}) => {
  const { mediaUrlService } = useContext(ServiceContext);
  const audioUrls = useMemo(
    () => items.map((item) => item.sound?.id && mediaUrlService.soundURL(item.sound.id)),
    [items, mediaUrlService]
  );
  const renderBody = () => {
    if (dictate) {
      return (
        <Player
          onCompleteAll={onAllItemsComplete}
          audioUrls={audioUrls}
        />
      );
    }
    return (
      <TextContent
        items={items}
        translationsVisible={translationsVisible}
        onScrollReachedEnd={onAllItemsComplete}
      />
    );
  };
  return (
    <StyledTrainerLayout>
      <ComprehensionTrainerTitle
        text={title || translations.trainerTitle[dictate ? 'audio' : 'text']}
      />
      <StyledWrapper>
        {description && (
          <StyledDescriptionWrapper>
            <ContextualInfo
              active={true}
              dataItemType="description"
              markupString={description}
            />
          </StyledDescriptionWrapper>
        )}
        <TrainerBody>
          {renderBody()}
        </TrainerBody>
      </StyledWrapper>
    </StyledTrainerLayout>
  );
};

ComprehensionTrainer.propTypes = {
  onAllItemsComplete: PropTypes.func.isRequired,
  translations: PropTypes.object,
  items: PropTypes.array,
  title: PropTypes.string,
  description: PropTypes.string,
  dictate: PropTypes.bool,
  translationsVisible: PropTypes.bool
};

export default withTranslations(getTranslations)(ComprehensionTrainer);
