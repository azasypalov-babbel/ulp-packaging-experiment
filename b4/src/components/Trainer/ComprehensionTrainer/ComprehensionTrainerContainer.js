import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';

import withInfoText from '../../shared/withInfoText';
import ContinueSheet from '../../ContinueSheet';
import { useSoundsPreload } from '../shared/useSoundsPreload';
import { QuestionWithChoicebuttonsInteraction } from '../../Interactions/QuestionWithChoicebuttons';
import { useSequence } from '../../shared/useSequence';
import * as sequenceActions from '../../../dux/sequence/actions';
import withNormalisedTrainerData from '../withNormalisedTrainerData';
import ComprehensionTrainer from './ComprehensionTrainer';
import TranslationVisibilityToggle from '../shared/TranslationVisibilityToggle';
import { useBackgroundColor } from '../../shared/hooks/useBackgroundColor';
import { useTheme } from '../../../providers/theme';

export const TRAINER_STATE = {
  READY: 'READY',
  DONE: 'DONE'
};

export const ComprehensionTrainerContainer = ({
  currentTrainerItemIndex: initialItemIndex,
  learnLanguageAlpha3,
  trainer,
  attemptItem,
  completeItem,
  onStart,
  onFinish
}) => {
  const contentItems = trainer.itemGroups[0]?.items || [];
  const questionItems = trainer.itemGroups[1]?.items || [];
  const initialQuestionIndex = initialItemIndex - contentItems.length;
  const {
    element: currentItem,
    next: nextItem,
    isLast: isLastItem,
    index: itemIndex
  } = useSequence(questionItems, initialQuestionIndex < 0 ? 0 : initialQuestionIndex);
  useSoundsPreload(contentItems);
  const theme = useTheme();
  useBackgroundColor(theme.cascada.iceGrayW50);
  const [state, setState] = useState(TRAINER_STATE.READY);
  const [translationsVisible, setTranslationsVisible] = useState(false);
  const [highlighted, setHightlighted] = useState(false);

  useEffect(() => {
    onStart({ scorableItemsCount: questionItems.length });
  }, [onStart, trainer, questionItems.length]);

  const handleAttempt = useCallback(
    ({ attempt }) => {
      const trainerData = {
        /* eslint-disable camelcase */
        interaction_mode: trainer.interaction,
        item_position_in_trainer: itemIndex,
        number_of_items_in_trainer: questionItems.length
        /* eslint-enable camelcase */
      };

      const trackingAttempt = {
        ...attempt,
        text: attempt.inputText,
        targetText: attempt.text
      };
      delete trackingAttempt.inputText;

      attemptItem(currentItem, trackingAttempt, trainerData);
    },
    [
      trainer,
      attemptItem,
      itemIndex,
      currentItem,
      questionItems.length
    ]
  );
  const handleItemComplete = useCallback(
    (mistakeCount) => {
      completeItem(currentItem, mistakeCount);
      if (isLastItem) setState(TRAINER_STATE.DONE);
      else nextItem();
    },
    [completeItem, currentItem, isLastItem, nextItem]
  );
  const handleAllItemsComplete = useCallback(() => {
    setHightlighted(true);
  }, []);

  return (
    <div
      data-trainer-type="comprehension"
      data-trainer-interaction={trainer.interaction}
      data-trainer-dictate={trainer.dictate}
    >
      {!trainer.dictate && trainer.translationVisibility !== 'none' && (
        <TranslationVisibilityToggle
          translationVisibility="partial"
          onClick={() => setTranslationsVisible((current) => !current)}
        />
      )}
      <ComprehensionTrainer
        items={contentItems}
        title={trainer.title}
        description={trainer.description}
        dictate={trainer.dictate}
        translationsVisible={translationsVisible}
        onAllItemsComplete={handleAllItemsComplete}
      />
      {state === TRAINER_STATE.READY && (
        <QuestionWithChoicebuttonsInteraction
          key={currentItem.id}
          item={currentItem}
          highlighted={highlighted}
          onAttempt={handleAttempt}
          onFinish={handleItemComplete}
          defaultOpened={contentItems.length === 0}
          learnLanguageAlpha3={learnLanguageAlpha3}
          amountOfQuestions={questionItems.length}
          currentQuestionOrder={itemIndex + 1}
        />
      )}
      {state === TRAINER_STATE.DONE && <ContinueSheet onClick={onFinish} />}
    </div>
  );
};

ComprehensionTrainerContainer.propTypes = {
  attemptItem: PropTypes.func.isRequired,
  completeItem: PropTypes.func.isRequired,
  currentTrainerItemIndex: PropTypes.number.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  onFinish: PropTypes.func.isRequired,
  onStart: PropTypes.func.isRequired,
  trainer: PropTypes.object.isRequired
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  locale: session.locale
});

const mapDispatchToProps = {
  skipItem: sequenceActions.skipItem,
  attemptItem: sequenceActions.attemptItem,
  completeItem: sequenceActions.completeItem
};

const wrapper = compose(
  withInfoText,
  withNormalisedTrainerData,
  connect(mapStateToProps, mapDispatchToProps)
);

export default wrapper(ComprehensionTrainerContainer);
