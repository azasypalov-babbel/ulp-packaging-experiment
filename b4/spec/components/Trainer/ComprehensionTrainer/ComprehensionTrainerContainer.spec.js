import React from 'react';
import { mount } from 'enzyme';

/* eslint-disable max-len */
import { ComprehensionTrainerContainer } from '../../../../src/components/Trainer/ComprehensionTrainer/ComprehensionTrainerContainer';
import { QuestionWithChoicebuttonsInteraction } from '../../../../src/components/Interactions/QuestionWithChoicebuttons';
/* eslint-enable max-len */

import { contentItems, questionsItems } from './defaultProps';
import { withServicesProvider } from '../../../../src/components/shared/withServices';
import withThemeProvider from '../../../../src/providers/theme';
import ContinueSheet from '../../../../src/components/ContinueSheet';
import ComprehensionTrainer from '../../../../src/components/Trainer/ComprehensionTrainer/ComprehensionTrainer';
import soundService from '../../../../src/services/soundService';
import TranslationVisibilityToggle from '../../../../src/components/Trainer/shared/TranslationVisibilityToggle';

jest.mock('../../../../src/components/Trainer/ComprehensionTrainer/ComprehensionTrainer', () => jest.fn(() => null));

jest.mock('../../../../src/services/soundService');
jest.mock('../../../../src/components/ContinueSheet', () => () => null);

jest.mock('../../../../src/components/Interactions/QuestionWithChoicebuttons', () => ({
  __esModule: true,
  QuestionWithChoicebuttonsInteraction: jest.fn(() => null)
}));

const defaultProps = {
  trainer: {
    interaction: 'choice',
    dictate: false,
    itemGroups: [
      { items: contentItems },
      { items: questionsItems }
    ]
  },
  onStart: jest.fn(),
  onFinish: jest.fn(),
  track: jest.fn(),
  learnLanguageAlpha3: 'DEU',
  currentTrainerItemIndex: 0,
  attemptItem: jest.fn(),
  completeItem: jest.fn(),
  clearInfoTextUI: jest.fn(),
  displayInfoText: jest.fn()
};

const Component = withServicesProvider(() => ({
  soundService,
  mediaUrlService: { soundURL: jest.fn((arg) => `${arg}.mp3`) }
}))(withThemeProvider(ComprehensionTrainerContainer));

const getWrapper = (props) => mount(
  <Component {...defaultProps} {...props} />
);

describe('ComprehensionTrainerContainer', () => {
  afterEach(() => {
    defaultProps.onStart.mockClear();
    defaultProps.onFinish.mockClear();
  });

  it('preloads sounds', () => {
    getWrapper();
    expect(soundService.preload.mock.calls).toMatchSnapshot();
  });

  it('triggers onStart on a new trainer', () => {
    const wrapper = getWrapper();
    expect(defaultProps.onStart).toBeCalledWith({ scorableItemsCount: 1 });
    wrapper.setProps({ forceUpdate: 1 });
    expect(defaultProps.onStart).toBeCalledTimes(1);
    wrapper.setProps({ trainer: { ...defaultProps.trainer } });
    expect(defaultProps.onStart).toBeCalledTimes(2);
  });

  it('renders ContinueSheet when all items are completed', () => {
    const wrapper = getWrapper();
    expect(wrapper.find(ContinueSheet).exists()).toBeFalsy();
    wrapper.find(QuestionWithChoicebuttonsInteraction).invoke('onFinish')(0);
    expect(wrapper.find(ContinueSheet).exists()).toBeTruthy();
  });

  it('triggers onFinish when ContinueSheet dismiss button clicked', () => {
    const wrapper = getWrapper();
    wrapper.find(QuestionWithChoicebuttonsInteraction).invoke('onFinish')(0);
    wrapper.find(ContinueSheet).invoke('onClick')();
    expect(defaultProps.onFinish).toBeCalledTimes(1);
  });

  it('highlights QuestionSheet expand button once all items are viewed or played', () => {
    const wrapper = getWrapper();
    expect(wrapper.find(QuestionWithChoicebuttonsInteraction).prop('highlighted')).toBeFalsy();
    wrapper.find(ComprehensionTrainer).invoke('onAllItemsComplete')();
    expect(wrapper.find(QuestionWithChoicebuttonsInteraction).prop('highlighted')).toBeTruthy();
  });

  it('has translation visibility toggle only on dictate flag set to false', () => {
    const wrapper = getWrapper();
    expect(wrapper.find(TranslationVisibilityToggle).exists()).toBeTruthy();
    wrapper.setProps({ trainer: { ...defaultProps.trainer, dictate: true } });
    expect(wrapper.find(TranslationVisibilityToggle).exists()).toBeFalsy();
  });

  it('passes the right visibility flag to ComprehensionTrainer component', () => {
    const wrapper = getWrapper();
    expect(wrapper.find(ComprehensionTrainer).prop('translationsVisible')).toBeFalsy();
    wrapper.find(TranslationVisibilityToggle).invoke('onClick')();
    expect(wrapper.find(ComprehensionTrainer).prop('translationsVisible')).toBeTruthy();
  });
});
