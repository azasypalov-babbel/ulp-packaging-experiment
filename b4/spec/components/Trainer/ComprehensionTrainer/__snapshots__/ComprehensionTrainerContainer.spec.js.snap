// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComprehensionTrainerContainer preloads sounds 1`] = `
Array [
  Array [
    "123.mp3",
  ],
  Array [
    "456.mp3",
  ],
  Array [
    "789.mp3",
  ],
]
`;
